fix: delete file if exist before uploading another one

Turn0xx opened the pull request:
-I couldn't do this with one query because MySql doesn't support RETURNING clause
-For retriveal tests I had to build a clean state for them to work because they were depending on Upload tests

Issue: #38 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

Turn0xx committed:
fix: delete file if exist before uploading another one

-I couldn't do this with one query because MySql doesn't support
RETURNING clause
-For retriveal tests I had to build a clean state for them to work
because they were depending on Upload tests

Turn0xx committed:
refactor: use deleteFile function in delete endpoint

jean-michelet commented on src/routes/api/tasks/index.ts:
```diff
@@ -360,4 +343,30 @@ function isErrnoException (error: unknown): error is NodeJS.ErrnoException {
   return error instanceof Error && 'code' in error
 }
 
+function buildFilePath (filename: string, fastify: FastifyInstance) {
+  return path.join(
+    import.meta.dirname,
+    '../../../..',
+    fastify.config.UPLOAD_DIRNAME,
+    fastify.config.UPLOAD_TASKS_DIRNAME,
+    filename
+  )
+}
+
+async function deleteFile (fileName: string, fastify: FastifyInstance) {
+  const filePath = buildFilePath(fileName, fastify)
+
+  try {
+    await fs.promises.unlink(filePath)
+  } catch (err) {
+    if (isErrnoException(err) && err.code === 'ENOENT') {
+      // A file could have been deleted by an external actor, e.g. system administrator.
+      // We log the error to keep a record of the failure, but consider that the operation was successful.
+      fastify.log.warn(`File path '${fileName}' not found`)
+    } else {
+      throw err
+    }
+  }
+}
```
I think we should create a new custom plugin for file operations .

We could use it a similar way:
```ts
fastify.fileHandler.uploadTaskImage(filename)
fastify.fileHandler.removeTaskImage(filename)

// Later
fastify.fileHandler.uploadUserImage(filename)
fastify.fileHandler.removeUserImage(filename)
```

Or:
```ts
fastify.fileHandler.upload(taskDirName, filename)
fastify.fileHandler.remove(taskDirName, filename)
```

Turn0xx committed:
feat: file handling plugin

jean-michelet commented on src/plugins/custom/file-handling.ts:
```diff
@@ -0,0 +1,76 @@
+declare module 'fastify' {
+  export interface FastifyInstance {
+    fileHandler: {
+      upload: (fileDir: FileDir, fileName: string, file: BusboyFileStream) => Promise<void>
+      remove: (fileDir: FileDir, fileName: string) => Promise<void>
```
```suggestion
      upload: typeof upload;
      remove: typeof remove;
```

Turn0xx replied:
i can't do this, because i have to bind manually the fastify instance in the decorate function

jean-michelet replied:
> i have to bind manually the fastify instance in the decorate function

Why?

jean-michelet commented on src/plugins/custom/file-handling.ts:
```diff
@@ -0,0 +1,76 @@
+  )
+}
+
+type FileDir = 'tasks'
+
+type FileDirConfigPath<T> = {
+  [K in keyof T]: K extends `UPLOAD_${string}_DIRNAME` ? K : never
+}[keyof T]
+
+type FileDirMap = {
+  [key in FileDir]: FileDirConfigPath<FileDirConfig>
+}
+
+const fileDirMap: FileDirMap = {
+  tasks: 'UPLOAD_TASKS_DIRNAME'
+}
```
This is unnecessary complex, just pass the path as a string argument.

jean-michelet commented on src/plugins/custom/file-handling.ts:
```diff
@@ -0,0 +1,76 @@
+
+export default fp(async (fastify) => {
+  fastify.decorate('fileHandler', {
+    upload: upload.bind(fastify),
+    remove: remove.bind(fastify)
```
This is not needed. 

Fastify instance (`this`) will be automatically passed if we use function declaration with `function` keyword (doesn't work with arrow functions).

Turn0xx replied:
I have to bind it manually, otherwise the instance passed is incorrect.. (empty object) 
I think that it is because of the way i use the `decorate` function, I passe to it an objects and not a function, so the functions inside of the object doesn't get the right instance.

jean-michelet replied:
> I passe to it an objects and not a function

Ha, ok I reviewed too quickly.

You can put all the functions you use directly in the registration callback so you get access to the fastify instance.
```ts
export default fp(async (fastify) => {
// do all the logic here
```

Turn0xx replied:
It's works, but i can't do this (upload and remove not anymore in scope)
```typescript
declare module 'fastify' {
  export interface FastifyInstance {
    fileHandler: {
      upload: typeof upload
      remove: typeof remove
    }
  }
}
```

jean-michelet commented on src/plugins/custom/file-handling.ts:
```diff
@@ -0,0 +1,76 @@
+  UPLOAD_TASKS_DIRNAME: string
+}
+
+function buildFilePath (fileDir: FileDir, filename: string, fastify: FastifyInstance) {
```
```suggestion
function buildFilePath (fastify: FastifyInstance, fileDir: FileDir, filename: string) {
```

jean-michelet commented on test/routes/api/tasks/tasks.test.ts:
```diff
@@ -622,6 +647,34 @@ describe('Tasks api (logged user only)', () => {
+
+        await app.knex<Task>('tasks').where({ id: taskId }).update({ filename: null })
+
+        const files = fs.readdirSync(uploadDirTask)
+        files.forEach((file) => {
+          const filePath = path.join(uploadDirTask, file)
+          fs.rmSync(filePath, { recursive: true })
+        })
```
You can create a function `removeFiles`, but the tests are run sequentially for db consistency, if call it in after hook, they will be removed before you reach this test suite.

Turn0xx replied:
> but the tests are run sequentially for db consistency, if called in after hook, they will be removed before you reach this test suite.

i didn't understand this

jean-michelet replied:
I mean that tests are not run in parralel, if you remove the files you uploaded during a test suite in `after` hook, it hsoudl be enough. You should not have to remove the files in `before` hook.

Turn0xx committed:
refactor(file-handling): change paramaters order

Turn0xx committed:
refactor(file-handling): explicit binding by factory & type interface by inference

Turn0xx committed:
chore: lint\

Turn0xx committed:
refactor(tasks test): clear dir function for reuse

Turn0xx committed:
chore: remove console.log

jean-michelet commented:
Install and run Sonarqube locally before pushing. They are also editor
extensions.

On Sat, Nov 2, 2024, 16:38 sonarcloud[bot] ***@***.***> wrote:

> [image: Quality Gate Failed]
> <https://sonarcloud.io/dashboard?id=fastify_demo&pullRequest=50> *Quality
> Gate failed*
>
> Failed conditions
>
> <https://camo.githubusercontent.com/bd1d5234b37410841ad8cd9d85c84eeecaec92ae18a95c926ae3a59a17c29a65/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6661696c65642d313670782e706e67> 15.1%
> Duplication on New Code
> <https://sonarcloud.io/component_measures?id=fastify_demo&pullRequest=50&metric=new_duplicated_lines_density&view=list>
> (required ≤ 3%)
>
> See analysis details on SonarCloud
> <https://sonarcloud.io/dashboard?id=fastify_demo&pullRequest=50>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fastify/demo/pull/50#issuecomment-2453025934>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A2J2327TMKPDVVCXZLW4KATZ6TWXVAVCNFSM6AAAAABQZ6NND6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINJTGAZDKOJTGQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Turn0xx committed:
refactor(tasks tests): move setup & tear down from child `describe` to parent `describe`

jean-michelet commented on src/plugins/custom/file-handling.ts:
```diff
@@ -0,0 +1,59 @@
+  )
+}
+
+function fileHandlerFactory (fastify: FastifyInstance) {
```
```suggestion
function createFileHandler (fastify: FastifyInstance) {
```

jean-michelet commented on src/routes/api/tasks/index.ts:
```diff
@@ -246,24 +244,21 @@ const plugin: FastifyPluginAsyncTypebox = async (fastify) => {
-        const affectedRows = await trx<Task>('tasks')
-          .where({ id })
-          .update({ filename })
+        const task = await trx<Task>('tasks').where({ id }).first()
```
```suggestion
        const task = await trx<Task>('tasks').select('filename').where({ id }).first()
```

jean-michelet commented on test/routes/api/tasks/tasks.test.ts:
```diff
@@ -509,6 +513,31 @@ describe('Tasks api (logged user only)', () => {
+      it('should delete existing image for a task before uploading a new one', async (t) => {
+        app = await build(t)
+
+        await uploadImageForTask(app, taskId, testImagePath, uploadDirTask)
+
+        const { mock: mockUnlink } = t.mock.method(fs.promises, 'unlink')
+
+        const form = new FormData()
+        form.append('file', fs.createReadStream(testImagePath))
+
+        const res = await app.injectWithLogin('basic', {
+          method: 'POST',
+          url: `/api/tasks/${taskId}/upload`,
+          payload: form,
+          headers: form.getHeaders()
+        })
+
+        assert.strictEqual(res.statusCode, 200)
+
+        const { message } = JSON.parse(res.payload)
+        assert.strictEqual(message, 'File uploaded successfully')
+
+        assert.strictEqual(mockUnlink.callCount(), 1)
```
A safer way to test this would be to upload the first time with a different image (so different filename):
```ts
await uploadImageForTask(
  app, 
  taskId, 
  imageToReplacePath,  // Here...
  uploadDirTask
 )
```

And check after the request that the file is not present in the upload dir anymore.

jean-michelet reviewed (approved):
Take a look at my feedbacks plz.

Besides that, LGTM!

Turn0xx committed:
refactor(tasks test)

- Generalize setup and tear down for all tasks image tests describe
blocks
- In replacing image upload test, assert that the old file is really
removed and the new file is persisted

Turn0xx committed:
fix: fileHandler type & lint

jean-michelet commented:
Can I merge @climba03003 @Fdawgs?

climba03003 reviewed (changes requested):
I would go for temporary file approach.
1. Save the file with random name.
2. Remove the old file.
3. Move the tmp file to proper location.

It can ensure when the transaction is rollback or file upload is failed.
You still have the copy of old file.

jean-michelet commented:
Do you plan to terminate this PR @Turn0xx? Or should I terminate it for you?

Turn0xx commented:
@jean-michelet I plan to finish it, but i don't have much time this weeks so i don't know when.

jean-michelet commented:
Fixed by #93
